raindrops-protocol/raindrops
# Namespace Program

Namespace Gatekeepers
- Namespace Gatekeepers are initialized by the program with `create_namespace_gatekeeper`, they start with 0 filters
- Filters are appended and removed via their respective instructions
- Currently the filter impl for Category and Key only use the namespace and mint fields respectively, what is the vision for these 2 filters?
- Does an artifact need to pass all filters to be included in the namespace or just 1?
- PermissivenessSettings are set at `initialize_namespace` but really are used with the Gatekeeper, I wonder if it makes more sense to set them at the gatekeeper level?


Artifact Caching
- An artifact must be added to a namespace via `join_namespace` to be eligible to be cached, an item cannot leave a namespace via `leave_namespace` without first being removed from the namespace cache
- There are 2 new fields in the `Namespace` Account, `max_pages` and `full_pages`, `max_pages` denotes how many pages a namespace is allowed to use for caching, this is currently hardcoded to 10. `full_pages` keeps track of which pages in the cache are full and cannot accept any more artifacts.
- The program will use `full_pages` to determine which page or index needs to be passed in via the client, this is so we always fill up the lower pages first and use all the space at our disposal. The client can query the artifact's account to figure out which page it's cached, this is useful to remove it from the cache as you will need to pass in the appropriate account.
- For an artifact to be joined and cached to a particular namespace, that artifact needs to expose 4 instructions for `join_namespace`, `leave_namespace`, `cache_artifact` and `uncache_artifact` For now I have just done `item_class` in the `item` program to test this out.

Client Side
- CLI stubbed out but not tested, SDK used by CLI and tests

Testing
- I need the keypairs for the programs to test in dev/main but I have setup local tests that run agains the validator simulation, these are runnable with `anchor test --skip-lint`
- Right now really only the happy paths are tested

Notes/Questions
- As we begin to do CPI and share types between programs, it may be a good idea to have another crate that is simply all the shareable types so that we dont get into dependency cycle issues. In addition we have a lot of similar helper functions in utils, these could be added there as well.
- Namespace Program as its written assumes no program which wants to be namespaced will bring in the Namespace program as a dependency as this creates a cycle.
- Seems like the raindrops typescript library uses an older version of @solana/spl-token, we should upgrade this

Next Steps
- If this PR and methods pertaining to caching and joining are acceptable, the next things would be to implement the instructions for the other crates and the rest of the item artifacts.
- More extensive testing to cover all the edge cases
- Add the necessary changes so that all programs pass anchor linting
- Add Vaults to accept $RAIN and other tokens as payment for joining namespaces
- CLI and test
